docs/rfcs/20220131_raft_snapshots.md

```diff
@@ -0,0 +1,150 @@
+ 
+ - Idempotency: Redpanda inspects the messages to update the list of ongoing sessions and then uses the list to reject the out-of-order requests.
+ - Offset translator: Redpanda tracks the reconfiguration batches to adjust Raft offsets into Kafka offsets.
+ - Transactions: Redpanda inspects the records to keep an eye on the ongoing transactions and maintain the right consistency level.
+ 
+At the same time, disks have limited capacity and data eventually loses actuality, so Redpanda supports multiple cleanup policies:
+ 
+ - Deletion
+ - Compaction
+ 
+The deletion clean up policy removes some messages (defined by age or size) from the head of the log.
+
+![](./imgs/20220131_log.png)
+
+The compaction clean up policy is a bit more complicated. Each record in Redpanda has a key and when compaction is enabled, Redpanda keeps a record at the tail of the log, but removes some or all of the records with the same key written before it.
+
+## Problem
+
+There is a problem with the integration of aggregate-dependent operations and the cleanup policies: Redpanda may remove data before the operation sees it. This is especially possible during reconfiguration because the new node has never seen a non-compacted log).
+
+![](./imgs/20220131_cleanup.png)
+
+For example, imagine that Redpanda stores withdraw and deposit history for a set of accounts and an aggregate-dependent operation needs an account balance.
+
+![](./imgs/20220131_compaction.png)
+
+We see that the state after compaction is completely different: `(2,-2,4)` vs `(3,1,4)`. The only way to guarantee the correct result is to cache the state before the compaction / deletion.
+
+Local caching isn't enough. Raft tolerates node failures as long as the majority are connected. This means a single node may fall behind the majority and behind the compaction horizon.
+
+![](./imgs/20220131_behind.png)
+
+When the isolated node rejoins the cluster it fetches an already-compacted log with holes, and the state diverges.
+
+# Reference-level explanation
+
+## Approach #1: Ship the snapshot (cached state) via the log
+
+The main idea is to use existing log replication to ship the snapshot. It's easier to implement, but just like the reconfiguration batches, it shifts the offsets and requires that we continue to rely on offset translation.
+
+### Overview
+
+When a leader decides to do compaction at offset X, it:
+ 
+  1. saves the current snapshot
+  2. waits until it's safe to do compaction:
+    a. segment containing X is closed
+    b. tx front is passed the end of the segment
+  3. replicates the snapshot
+ 
+When a leader or a follower sees the snapshot via `state_machine::apply` (this only happens when the batch is replicated to the majority and can't be lost), it:
+
+  - performs the compaction when the current node is the leader and it sent the snapshot (within the current term)
+  - when the snapshot isn't applied:
```

Comment:
I don't quite understand.  Maybe this line shouldn't be indented under "when a leader of follower sees the snapshot via apply"?
 Can you elaborate on this scenario? Is the snapshot not applied because leadership changed?